server/auth_callout.go

```diff
@@ -0,0 +1,329 @@
+		if err := c.RegisterNkeyUser(nkuser); err != nil {
+			s.Warnf("Could not register auth callout user: %v", err)
+			respCh <- false
+			return
+		}
+
+		// See if the response wants to override the userID.
+		if juc.Name != _EMPTY_ {
+			c.mu.Lock()
+			c.opts.Username = juc.Name
+			// Clear any others.
+			c.opts.Nkey = _EMPTY_
+			c.pubKey = _EMPTY_
+			c.opts.Token = _EMPTY_
+			c.mu.Unlock()
+		}
+
+		// Check if we need to set an auth timer if the user jwt expires.
+		c.setExpiration(juc.Claims(), validFor)
+
+		respCh <- true
+	}
+
+	sub, err := acc.subscribeInternal(reply, processReply)
+	if err != nil {
+		s.Warnf("Error setting up reply subscription for auth request: %v", err)
+		return false
+	}
+	defer acc.unsubscribeInternal(sub)
+
+	// Build our request claims.
+	claim := jwt.NewAuthorizationRequestClaims(AuthRequestSubject)
+	// Set expected public user nkey.
+	claim.UserNkey = pub
+
+	s.mu.RLock()
+	claim.Server = jwt.ServerID{
+		Name: s.info.Name,
+		Host: s.info.Host,
+		ID:   s.info.ID,
+	}
+	// Check if we have been requested to encrypt.
+	if xkp != nil {
+		claim.Server.XKey = xkey
+	}
+	s.mu.RUnlock()
+
+	authTimeout := secondsToDuration(s.getOpts().AuthTimeout)
+	claim.Expires = time.Now().Add(time.Duration(authTimeout)).UTC().Unix()
+	if opts.AuthCallout != nil {
+		claim.Audience = opts.AuthCallout.Issuer
+	} else {
+		claim.Audience = acc.Name
+	}
+
+	// Grab client info for the request.
+	c.mu.Lock()
+	c.fillClientInfo(&claim.ClientInformation)
+	c.fillConnectOpts(&claim.ConnectOptions)
+	// If we have a sig in the client opts, fill in nonce.
+	if claim.ConnectOptions.SignedNonce != _EMPTY_ {
+		claim.ClientInformation.Nonce = string(c.nonce)
+	}
+
+	// TLS
+	if c.flags.isSet(handshakeComplete) && c.nc != nil {
+		var ct jwt.ClientTLS
+		conn := c.nc.(*tls.Conn)
+		cs := conn.ConnectionState()
+		ct.Version = tlsVersion(cs.Version)
+		ct.Cipher = tlsCipher(cs.CipherSuite)
+		// Check verified chains.
+		for _, vs := range cs.VerifiedChains {
+			var certs []string
+			for _, c := range vs {
+				blk := &pem.Block{
+					Type:  "CERTIFICATE",
+					Bytes: c.Raw,
+				}
+				certs = append(certs, string(pem.EncodeToMemory(blk)))
+			}
+			ct.VerifiedChains = append(ct.VerifiedChains, certs)
+		}
+		// If we do not have verified chains put in peer certs.
+		if len(ct.VerifiedChains) == 0 {
+			for _, c := range cs.PeerCertificates {
+				blk := &pem.Block{
+					Type:  "CERTIFICATE",
+					Bytes: c.Raw,
+				}
+				ct.Certs = append(ct.Certs, string(pem.EncodeToMemory(blk)))
+			}
+		}
+		claim.TLS = &ct
+	}
+	c.mu.Unlock()
+
+	b, err := claim.Encode(s.kp)
+	if err != nil {
+		s.Warnf("Error encoding auth request claim: %v", err)
+		return false
+	}
+	req := []byte(b)
+	var hdr map[string]string
+
+	// Check if we have been asked to encrypt.
+	if xkp != nil {
+		req, err = xkp.Seal([]byte(req), pubAccXKey)
+		if err != nil {
+			s.Warnf("Error encrypting auth request claim: %v", err)
```

Comment:
Since this might vary based on the account information, should this log the account for which we tried this?  And should this be an account advisory in operator mode?

Reply:
Will change, see above on advisories and rate limiting for server warnings etc.